### PR TITLE
Configurable Download Command [WIP]

### PR DESCRIPTION
**Pull request by malnick**
## High Level Description
This PR makes the download command for `dcos-dowload.service` configurable for either `curl` (default) or `aws s3`. This PR adds a calc function for determining the command to use based on the URL protocol. If the protocol evaluates to `s3://`, it uses the `aws s3 cp` command. If it's anything else, it defaults to the standard `curl` command. 

This enables users who are on AWS to push and pull the bootstrap of a given configuration to S3. This in turn gives a highly available location of their configuration, which is useful if their clusters leverage autoscaling groups. Without a HA location for their bootstrap contents, ASG becomes less durable. 

### Questions & Caveats:
1. This currently doesn't pass a fully qualified path to `aws` CLI, how should this be determined? Should this be shipped in `/opt/mesosphere/bin` or should it be another parameter in the config.yaml for the user to configure? Or, should we do something like `$(which aws) s3 ...`? 

2. This implementation assumes the S3 bucket and prefix have completely open permissions (public) or that the EC2 instance profile allows access to this resource. It does not assume any local, explicit API key access.

## Checklist for all PR's

  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [ ] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [ ] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

**Comment by malnick** [comment]
@mesosphere-mergebot label Request For Comment

**Comment by spahl** [comment]
I don't think we need to do some interpretation and try to use `aws` cli when seeing s3 but rather make it simpler.

Let the user specify the full download command in addition to the bootstrap url. That way is not only simpler code but also more flexible (yes even torrent:-P).

This will also need a new test of course.

**Comment by malnick** [comment]
Sounds good to me. My first pass on this was exactly that, but after looking at the code, I realized the tmp dir is mandatory since the installer must expect that. If we let the user define the entire command, we need to make sure they don't shoot themselves in the foot with that caveat, and document the fact that they need to place the download stuff into the tmp dir where the installer expects it.

File: gen/calc.py [hunk 1]
```diff
@@ -133,6 +133,17 @@ def validate_ip_port_list(json_str: str):
     validate_ipv4_addresses(ip_list)
 
 
+def calculate_bootstrap_download_command(bootstrap_tmp_dir, bootstrap_url, bootstrap_id):
+    if 's3://' in bootstrap_url:
```
Comment (vespian) [inline gen/calc.py]: `.startswith()`?

**Comment by malnick** [comment]
I'm only able to see the code quality tests in TC after making an account. Is there public access to the other integration tests? https://www.dropbox.com/s/bjs64nocv4szja1/Screenshot%202017-07-31%2012.46.22.png?dl=0

**Pull request closed without merging**